src/main/java/com/facebook/openwifirrm/modules/ModelerUtils.java

```diff
@@ -202,4 +212,110 @@ public static double calculateTPCMetrics(
+	 * Pre-802.11n, if two entries with the same bssid are in the same channel (and
+	 * therefore frequency, since they share a one-to-one mapping), that is enough
+	 * to aggregate them. However, 802.11n introduced channel bonding, so for
+	 * 802.11n onwards, channel width must be checked.
+	 *
+	 * @param entry1
+	 * @param entry2
+	 * @return true if the entries should be aggregated
+	 */
+	private static boolean matchesForAggregation(WifiScanEntry entry1, WifiScanEntry entry2) {
+		// do not check SSID (other SSIDs can contribute to interference, and SSIDs can
+		// change any time)
+		return Objects.equals(entry1.bssid, entry2.bssid) && entry1.frequency == entry2.frequency
+				&& entry1.channel == entry2.channel
+				&& HTOperationElement.matchesHtForAggregation(entry1.ht_oper, entry2.ht_oper)
+				&& VHTOperationElement.matchesVhtForAggregation(entry1.vht_oper, entry2.vht_oper);
+	}
+
+	/**
+	 * For each AP, for each other AP that sent a wifiscan entry to that AP, this
+	 * method calculates an aggregate wifiscan entry with an aggregated RSSI.
+	 *
+	 * @param dataModel          the data model which includes the latest wifiscan
+	 *                           entries
+	 * @param obsoletionPeriodMs for each (scanning AP, responding AP) tuple, the
+	 *                           maximum amount of time (in milliseconds) it is
+	 *                           worth aggregating over, starting from the most
+	 *                           recent scan entry for that tuple, and working
+	 *                           backwards in time. An entry exactly
+	 *                           {@code obsoletionPeriodMs} ms earlier than the most
+	 *                           recent entry is considered non-obsolete (i.e., the
+	 *                           "non-obsolete" window is inclusive). Must be
+	 *                           non-negative.
+	 * @param agg                an aggregator to calculate the aggregated RSSI
+	 *                           given recent wifiscan entries' RSSIs.
+	 * @return a map from AP serial number to a map from BSSID to an "aggregated
+	 *         wifiscan entry" (i.e., a single entry with its {@code signal}
+	 *         attribute modified to be the aggregated signal value instead of the
+	 *         value in just the most recent entry for that (AP serial number,
+	 *         BSSID) tuple. The returned map will only contain APs which received
+	 *         at least one non-obsolete wifiscan entry from a BSS.
+	 */
+	public static Map<String, Map<String, WifiScanEntry>> getAggregatedWifiScans(Modeler.DataModel dataModel,
+			long obsoletionPeriodMs,
+			Aggregator<Double> agg) {
+		if (obsoletionPeriodMs < 0) {
+			throw new IllegalArgumentException("obsoletionPeriodMs must be non-negative.");
+		}
+		Map<String, Map<String, WifiScanEntry>> aggregatedWifiScans = new HashMap<>();
+		for (Map.Entry<String, List<List<WifiScanEntry>>> apToScansMapEntry : dataModel.latestWifiScans.entrySet()) {
+			String ap = apToScansMapEntry.getKey();
```

Comment:
nit: we typically are calling this `serialNumber`, use same naming for consistency